### PR TITLE
Path mappings

### LINKED ISSUE #54: Mapping more than one path for source code
I was wondering if it's possible to map more than one path in vscode?
I'm working with Vagrant to develop a WordPress theme, and I'm using two shared folders on the guest.
I'm sharing one folder for the root of WordPress installation called `www` and the other one which is in the same path called `main`, shared on the guest within `themes` folder of WordPress installation.

**Vagrantfile:**

```
config.vm.synced_folder "./www/", "/var/www", :mount_options => ["dmode=777", "fmode=666"]
config.vm.synced_folder "./main", "/var/www/wp-content/themes/main", :mount_options => ["dmode=777", "fmode=666"]
```

The problem is with the current launch.json file I'm not able to define two paths for my source codes, and when I'm debugging, VSCode doesn't hit the breakpoints anywhere in the `main` folder!

**Take a look at the hierarchy below:**

Folder Hierarchy

It's worth mentioning I'm currently achieving this using Sublime Text, but I'd like to work in VSCode instead.

**Sublime Text project settings:**

```
{
    "settings": {
        "xdebug": {
            "path_mapping": {
                "/var/www/wp-content/themes/main" : "/Users/[user]/Documents/[project]/main",
                "/var/www" : "/Users/[user]/Documents/[project]/www",
             },
             "super_globals": true,
             "close_on_stop": true
        }
    }
}
```

**Extra info:**

PHP version: 5.6.14-1
XDebug version: 2.3.3

Your launch.json: 

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Listen for XDebug",
            "type": "php",
            "request": "launch",
            "port": 9000,
            "log": true,
            "serverSourceRoot": "/var/www",
            "localSourceRoot": "/Users/[user]/Documents/[project]/www"
        },
        {
            "name": "Launch currently open script",
            "type": "php",
            "request": "launch",
            "program": "${file}",
            "cwd": "${fileDirname}",
            "port": 9000
        }
    ]
}
```

Thanks in advance.

Amin

### PR DESCRIPTION
**Pull request by johnrom**
Added Path Mapping Support via `pathMappings` option. Added Tests. Reorganized test project to support path mappings.

fixes #54

### CODE REVIEW

File: src/paths.ts [hunk 1]
```diff
@@ -5,34 +5,87 @@ import * as url from 'url';
...
     let serverPath = decodeURI(fileUri.pathname!);
+
     // strip the trailing slash from Windows paths (indicated by a drive letter with a colon)
     const serverIsWindows = /^\/[a-zA-Z]:\//.test(serverPath);
+
     if (serverIsWindows) {
         serverPath = serverPath.substr(1);
     }
+
+    if ( pathMapping !== {} && typeof pathMapping !== 'undefined' ) {
```
Comment (aminpaks) [inline src/paths.ts]: Object references are never the same thus `{} !== {}` no?

Comment (johnrom) [inline src/paths.ts]: hm, you're correct.

Comment (aminpaks) [inline src/paths.ts]: Just check for own properties by `Object.keys` maybe?

File: src/paths.ts [hunk 2]
```diff
@@ -6,34 +6,95 @@ import * as path from 'path';
 import {decode} from 'urlencode';
 
 /** converts a server-side XDebug file URI to a local path for VS Code with respect to source root settings */
-export function convertDebuggerPathToClient(fileUri: string|url.Url, localSourceRoot?: string, serverSourceRoot?: string): string {
+export function convertDebuggerPathToClient(fileUri: string|url.Url, localSourceRoot?: string, serverSourceRoot?: string, pathMapping?: { [index: string]: string; }): string {
```
Comment (felixfbecker) [inline src/paths.ts]: isn't `localSourceRoot` and `serverSourceRoot` obsolete with `pathMappings`?

File: src/paths.ts (hunk 1 above)
Comment (johnrom) [inline src/paths.ts]: removed in d0052265ab203e67b8a38c7acd4731b6393471af

File: src/paths.ts [hunk 3]
```diff
@@ -6,34 +6,95 @@ import * as path from 'path';
...
     if (typeof fileUri === 'string') {
         fileUri = url.parse(fileUri);
     }
+
     // convert the file URI to a path
     let serverPath = decode(fileUri.pathname!);
+
     // strip the trailing slash from Windows paths (indicated by a drive letter with a colon)
     const serverIsWindows = /^\/[a-zA-Z]:\//.test(serverPath);
+
```
Comment (felixfbecker) [inline src/paths.ts]: please don't try to modify the existing code style.

File: src/paths.ts [hunk 4]
```diff
@@ -6,34 +6,95 @@ import * as path from 'path';
...
     if (serverIsWindows) {
         serverPath = serverPath.substr(1);
     }
+
+    if ( typeof pathMapping !== 'undefined' ) {
```
Comment (felixfbecker) [inline src/paths.ts]: please use the existing code style (no spaces inside parens)

File: src/paths.ts [hunk 5]
```diff
@@ -6,34 +6,95 @@ import * as path from 'path';
...
+        let mappedServerSource: string;
+
+        for (mappedServerSource of Object.keys(pathMapping) ) {
+            let mappedLocalSource: string = pathMapping[mappedServerSource];
+
+            // normalize slashes for windows-to-unix
+            // I thought normalize did this automagically..
+            if ( process.platform !== 'win32' ) {
+                mappedServerSource = mappedServerSource.replace(/\\/g, '/');
```
Comment (felixfbecker) [inline src/paths.ts]: what if the server is Windows?

File: src/test/paths.ts [hunk 6]
```diff
@@ -3,6 +3,64 @@ import {isSameUri, convertClientPathToDebugger, convertDebuggerPathToClient} fro
 import * as assert from 'assert';
 
 describe('paths', () => {
+    let undef: string;
+
+    const pathMapping = {
```
Comment (felixfbecker) [inline src/test/paths.ts]: I prefer to define the test cases through mocha's `it` DSL over having a config object, it's a lot easier to debug if the values are always inside the test

File: src/test/paths.ts [hunk 7]
```diff
@@ -56,6 +114,43 @@ describe('paths', () => {
...
         });
+
+        describe('with path mapping', () => {
+            let sources = pathMapping.local;
+            let results = pathMapping.remote;
+
+            // unix to unix
+            it('should convert a unix path to a unix URI', () => {
+                // site
+                assert.equal( convertClientPathToDebugger(sources.unix.site, undef, undef, pathMapping.unixToUnix), results.unix.site );
```
Comment (felixfbecker) [inline src/test/paths.ts]: why use `undef` and not `undefined`?

File: src/test/adapter.ts [hunk 8]
```diff
@@ -9,7 +9,7 @@ const assert = chai.assert;
 
 describe('PHP Debug Adapter', () => {
 
-    const TEST_PROJECT = path.normalize(__dirname + '/../../testproject');
+    const TEST_PROJECT = path.normalize(__dirname + '/../../testproject/main');
```
Comment (felixfbecker) [inline src/test/adapter.ts]: why is this needed?

File: src/paths.ts (hunk 2 above)
Comment (johnrom) [inline src/paths.ts]: Users that update will use "localSourceRoot" and "serverSourceRoot". It was easier to be backwards compatible with an optional extra argument, but I can certainly reformat it to inject localSourceRoot and serverSourceRoot into the config object if they exist.

(this is my first open source node contribution so I apologize for any code style differences)

File: src/paths.ts (hunk 4 above)
Comment (aminpaks) [inline src/paths.ts]: No linting!?

File: src/test/adapter.ts (hunk 8 above)
Comment (johnrom) [inline src/test/adapter.ts]: Some tests use /variables.php which has been moved to /testproject/main/variables.php to test multiple source roots.

File: src/paths.ts (hunk 2 above)
Comment (felixfbecker) [inline src/paths.ts]: BC is great, but this function should only take one source of truth for mappings

File: src/paths.ts (hunk 4 above)
Comment (felixfbecker) [inline src/paths.ts]: There is linting, but this isn't caught.

Comment (aminpaks) [inline src/paths.ts]: Maybe turn this on => [Whitespace](https://palantir.github.io/tslint/rules/whitespace/)

Comment (felixfbecker) [inline src/paths.ts]: it is on

File: src/paths.ts (hunk 5 above)
Comment (johnrom) [inline src/paths.ts]: This handles the case where local is Linux and server is Windows. This should be the only time backslashes need be converted because local windows will normalize the backslashes in path.relative just fine. Linux just sees the backslashes as part of the filename so path.relative returns the wrong value.

File: src/test/paths.ts (hunk 7 above)
Comment (johnrom) [inline src/test/paths.ts]: ... Dunno.

File: src/test/adapter.ts (hunk 8 above)
Comment (felixfbecker) [inline src/test/adapter.ts]: but I don't see any new tests that rely on this?

File: src/paths.ts [hunk 9]
```diff
@@ -6,34 +6,95 @@ import * as path from 'path';
...
+        for (mappedServerSource of Object.keys(pathMapping) ) {
+            let mappedLocalSource: string = pathMapping[mappedServerSource];
+
+            let localRelative: string = path.relative(mappedLocalSource, localPath);
+
+            if ( localRelative && localRelative.length > 0 ) {
+                let relative: number = localRelative.indexOf('..');
+
+                // if does not start with ..
+                if ( relative !== 0 ) {
```
Comment (felixfbecker) [inline src/paths.ts]: this is not used anywhere else, so no reason to save it in a variable. You can also use `.includes()` if I'm not mistaken.

File: src/paths.ts [hunk 10]
```diff
@@ -6,34 +6,95 @@ import * as path from 'path';
...
+export function convertClientPathToDebugger(localPath: string, localSourceRoot?: string, serverSourceRoot?: string, pathMapping?: { [index: string]: string; }): string {
+
     let localFileUri = fileUrl(localPath, {resolve: false});
     let serverFileUri: string;
-    if (serverSourceRoot && localSourceRoot) {
+
+    if ( typeof pathMapping !== 'undefined' ) {
+        let mappedServerSource: string;
+
+        for (mappedServerSource of Object.keys(pathMapping) ) {
```
Comment (felixfbecker) [inline src/paths.ts]: you should declare the `let` here so it's scoped to the loop body

File: src/test/adapter.ts (hunk 8 above)
Comment (johnrom) [inline src/test/adapter.ts]: I'm assuming that you're requesting a test that involves `/testproject/secondary`? Please give me a little direction on this.

Comment (felixfbecker) [inline src/test/adapter.ts]: I'm fine without a test like that given that `tests/paths.ts` unit-tests the path mapping, but I don't understand why all these files have to be moved then if they are essentially not used at all.

Comment (johnrom) [inline src/test/adapter.ts]: when testing development of this feature, I had to test that multiple paths were working, so I created multiple paths in `testproject`. I thought that in the future other developers would also need to test multiple paths, but if you'd like I can move the files back.

Comment (felixfbecker) [inline src/test/adapter.ts]: The testproject should only contain fixtures used by (automatic) tests

**Comment by johnrom** [comment]
I believe I've made all requested changes. Let me know if you need anything else.

File: src/paths.ts [hunk 11]
```diff
@@ -6,7 +6,9 @@ import * as path from 'path';
 import {decode} from 'urlencode';
 
 /** converts a server-side XDebug file URI to a local path for VS Code with respect to source root settings */
-export function convertDebuggerPathToClient(fileUri: string|url.Url, localSourceRoot?: string, serverSourceRoot?: string): string {
+export function convertDebuggerPathToClient(fileUri: string|url.Url, pathMapping?: { [index: string]: string; }): string {
+    let localSourceRoot: string = '';
+    let serverSourceRoot: string = '';
```
Comment (felixfbecker) [inline src/paths.ts]: why initialize with an empty string?

File: src/paths.ts [hunk 12]
```diff
@@ -17,8 +19,23 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
     if (serverIsWindows) {
         serverPath = serverPath.substr(1);
     }
+    if (typeof pathMapping !== 'undefined') {
```
Comment (felixfbecker) [inline src/paths.ts]: shorten to `if (pathMapping)`

File: src/paths.ts [hunk 13]
```diff
@@ -17,8 +19,23 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
...
+        for (let mappedServerSource of Object.keys(pathMapping) ) {
```
Comment (felixfbecker) [inline src/paths.ts]: `const`, I would name this `mappedServerPath`

File: src/paths.ts [hunk 14]
```diff
@@ -17,8 +19,23 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
...
+            let mappedLocalSource: string = pathMapping[mappedServerSource];
```
Comment (felixfbecker) [inline src/paths.ts]: don't annotate types that can be inferred

File: src/paths.ts [hunk 15]
```diff
@@ -17,8 +19,23 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
...
+            if (serverRelative.length > 0 && serverRelative.indexOf('..') !== 0) {
+                serverSourceRoot = mappedServerSource;
+                localSourceRoot = mappedLocalSource;
+                break;
+            }
+        }
+    }
     let localPath: string;
-    if (serverSourceRoot && localSourceRoot) {
+    if (serverSourceRoot !== '' && localSourceRoot !== '') {
```
Comment (felixfbecker) [inline src/paths.ts]: this change is unneeded

File: src/paths.ts [hunk 16]
```diff
@@ -30,10 +47,24 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
...
 
 /** converts a local path from VS Code to a server-side XDebug file URI with respect to source root settings */
-export function convertClientPathToDebugger(localPath: string, localSourceRoot?: string, serverSourceRoot?: string): string {
+export function convertClientPathToDebugger(localPath: string, pathMapping?: { [index: string]: string; }): string {
+    let localSourceRoot: string = '';
+    let serverSourceRoot: string = '';
     let localFileUri = fileUrl(localPath, {resolve: false});
     let serverFileUri: string;
-    if (serverSourceRoot && localSourceRoot) {
+    if ( typeof pathMapping !== 'undefined' ) {
```
Comment (felixfbecker) [inline src/paths.ts]: remove spaces, same comment as above

File: src/paths.ts [hunk 17]
```diff
@@ -30,10 +47,24 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
 }
 
 /** converts a local path from VS Code to a server-side XDebug file URI with respect to source root settings */
-export function convertClientPathToDebugger(localPath: string, localSourceRoot?: string, serverSourceRoot?: string): string {
+export function convertClientPathToDebugger(localPath: string, pathMapping?: { [index: string]: string; }): string {
+    let localSourceRoot: string = '';
+    let serverSourceRoot: string = '';
```
Comment (felixfbecker) [inline src/paths.ts]: see above

File: src/paths.ts [hunk 18]
```diff
@@ -30,10 +47,24 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
...
     let localFileUri = fileUrl(localPath, {resolve: false});
     let serverFileUri: string;
-    if (serverSourceRoot && localSourceRoot) {
+    if ( typeof pathMapping !== 'undefined' ) {
+        let mappedServerSource: string;
```
Comment (felixfbecker) [inline src/paths.ts]: declare inside the loop

File: src/paths.ts [hunk 19]
```diff
@@ -30,10 +47,24 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
...
+        for (mappedServerSource of Object.keys(pathMapping) ) {
+            let mappedLocalSource: string = pathMapping[mappedServerSource];
+            let localRelative: string = path.relative(mappedLocalSource, localPath);
```
Comment (felixfbecker) [inline src/paths.ts]: remove inferrable types

File: src/paths.ts [hunk 20]
```diff
@@ -30,10 +47,24 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
...
+            if (localRelative.length > 0 && localRelative.indexOf('..') !== 0) {
+                serverSourceRoot = mappedServerSource;
+                localSourceRoot = mappedLocalSource;
+                break;
+            }
+        }
+    }
+    if (serverSourceRoot !== '' && localSourceRoot !== '') {
```
Comment (felixfbecker) [inline src/paths.ts]: same as above

File: package.json [hunk 21]
```diff
@@ -149,6 +149,11 @@
                 "type": "string",
                 "description": "The source root on this machine that is the equivalent to the serverSourceRoot on the server."
               },
+              "pathMappings": {
+                "type": "object",
+                "default": {},
+                "description": "A mapping of multiple localSourceRoots on this machine to remote serverSourceRoots on the server."
```
Comment (felixfbecker) [inline package.json]: this shouldn't mention the name `localSourceRoot` anymore, and the old settings should be deprecated to not have two settings doing the same thing without clear precedence

File: src/phpDebug.ts [hunk 22]
```diff
@@ -180,6 +182,14 @@ class PhpDebugSession extends vscode.DebugSession {
     }
 
     protected async launchRequest(response: VSCodeDebugProtocol.LaunchResponse, args: LaunchRequestArguments) {
+        if ( args.hasOwnProperty('localSourceRoot') && typeof args.localSourceRoot !== 'undefined' && args.hasOwnProperty('serverSourceRoot') && typeof args.serverSourceRoot !== 'undefined') {
```
Comment (felixfbecker) [inline src/phpDebug.ts]: shorten to `if (args.localSourceRoot && args.serverSourceRoot)`

File: src/phpDebug.ts [hunk 23]
```diff
@@ -180,6 +182,14 @@ class PhpDebugSession extends vscode.DebugSession {
...
+            let pathMappings: {[index: string]: string} = {};
+            if ( args.hasOwnProperty('pathMappings') && typeof args.pathMappings !== 'undefined') {
```
Comment (felixfbecker) [inline src/phpDebug.ts]: same

File: src/paths.ts [hunk 24]
```diff
@@ -17,8 +19,23 @@ export function convertDebuggerPathToClient(fileUri: string|url.Url, localSource
     if (serverIsWindows) {
         serverPath = serverPath.substr(1);
     }
+    if (typeof pathMapping !== 'undefined') {
+        for (let mappedServerSource of Object.keys(pathMapping) ) {
+            let mappedLocalSource: string = pathMapping[mappedServerSource];
+            // normalize slashes for windows-to-unix
+            if (process.platform !== 'win32') {
```
Comment (felixfbecker) [inline src/paths.ts]: I don't think this check is needed, this should be a pure function

File: src/test/paths.ts [hunk 25]
```diff
@@ -23,37 +23,57 @@ describe('paths', () => {
...
             });
-            it('should convert a unix path to a windows URI', () => {
-                const localSourceRoot = '/home/felix/myproject';
-                const serverSourceRoot = 'C:\\Program Files\\Apache\\2.4\\htdocs';
-                assert.equal(
-                    convertClientPathToDebugger('/home/felix/myproject/test.php', localSourceRoot, serverSourceRoot),
-                    'file:///C:/Program%20Files/Apache/2.4/htdocs/test.php'
-                );
+            // unix to windows
+            (process.platform === 'win32' ? it : it.skip)('should convert a unix path to a windows URI', () => {
```
Comment (felixfbecker) [inline src/test/paths.ts]: if the platform check is removed, these should always run

File: testproject/.vscode/launch.json [hunk 26]
```diff
@@ -7,15 +7,21 @@
...
       "name": "Launch",
       "request": "launch",
       "type": "php",
       "program": "${file}",
       "cwd": "${workspaceRoot}",
-      "externalConsole": false
+      "externalConsole": false,
+      "pathMappings": {
+        "/var/www/html": "${workspaceRoot}"
+      }
```
Comment (felixfbecker) [inline testproject/.vscode/launch.json]: this setting shouldn't be the default here

File: package.json (hunk 21 above)
Comment (felixfbecker) [inline package.json]: The description says this map is local -> server, but testproject/launch.json looks like the opposite

Comment (johnrom) [inline package.json]: I'm not sure what you mean by "the old settings should be deprecated". I'm assuming I should just add "Deprecated: " to the beginning of the description?

File: src/paths.ts (hunk 13 above)
Comment (johnrom) [inline src/paths.ts]: mappedServerSource (or mappedServerPath) is modified by the slash normalization below

File: src/paths.ts (hunk 11 above)
Comment (johnrom) [inline src/paths.ts]: if these aren't initialized, then I receive an error testing for `serverSourceRoot && localSourceRoot` "variable is used before it is assigned"

File: package.json (hunk 21 above)
Comment (felixfbecker) [inline package.json]: Yep, and iirc you could also set `deprecated: true`

File: src/paths.ts (hunk 11 above)
Comment (felixfbecker) [inline src/paths.ts]: You need to type them as `string | undefined`

**Comment by johnrom** [comment]
FYI, I believe I've made all requested changes here.

File: src/paths.ts (hunk 4 above)
Comment (aminpaks) [inline src/paths.ts]: Just a friendly suggestion: tslint now has the [space within parens](https://github.com/palantir/tslint/commit/524c176684322facdb859397f9c457cc602f3349), you may enable it.

**Comment by johnrom** [comment]
FYI, I believe I've made all requested changes here. Let me know if there's anything else, or at least remove the "requested changes" attribute if they have been made properly -- if that's possible.

**Comment by lbod** [comment]
@johnrom thanks for this. Works a treat with my first attempt at debugging in vscode with docker wordpress/xdebug. It'll be vscode itself but the debugger response into mapped paths is very fast compared to atom.
